functional-programming/src/fp/utils.ts

```diff
@@ -0,0 +1,55 @@
+import { Maybe, none, some } from "./maybe";
+
+export const constant = <A>(a: A) => () => a;
+
+/**
+ * Performs left-to-right function composition
+ */
+export function flow<A, B, C>(fa: (a: A) => B, fb: (b: B) => C): (a: A) => C;
+export function flow<A, B, C, D>(fa: (a: A) => B, fb: (b: B) => C, fc: (c: C) => D): (a: A) => D;
+export function flow<A, B, C, D, E>(fa: (a: A) => B, fb: (b: B) => C, fc: (c: C) => D, fd: (d: D) => E): (a: A) => E;
+export function flow(...fns: Array<(...args: Array<string>) => string>) {
+  return (a: string) => fns.reduce(
+    (acc, fn) => fn(acc),
+    a
+  );
+}
+
+/**
+ * Pipes the value into the pipeline of functions
+ * Handy for automatic data typing
+ */
+export function pipe<A, B>(a: A, fb: (a: A) => B): B;
+export function pipe<A, B, C>(a: A, fb: (a: A) => B, fc: (b: B) => C): C;
+export function pipe<A, B, C, D>(a: A, fb: (a: A) => B, fc: (b: B) => C, fd: (c: C) => D): D;
+export function pipe<A, B, C, D, E>(a: A, fb: (a: A) => B, fc: (b: B) => C, fd: (c: C) => D, fe: (d: D) => E): E;
+export function pipe(a: string, ...fns: Array<(...args: Array<string>) => string>) {
+  return fns.reduce(
+    (acc, fn) => fn(acc),
+    a
+  );
+}
```

Comment:
```
type PipeFunction<T> = (arg: T) => T;

export function pipe(a: string, ...fns: Array<PipeFunction<string>>): string {
 return fns.reduce((acc, fn) => fn(acc), a);
}
```
can we write in such manner? I'm not super hero of typescript, but do we really need to describe type as sequence execution??